### PR TITLE
Documentation Panel

### LINKED ISSUE #7924: Documentation in new Component Browser
This task is about porting the old HTML documentation generator to the new GUI.

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

Implementing most parts of the Documentation Panel in the new GUI.

Known issues:
- Links do not work (yet, covered by #7992)
- Some pages are entirely empty – I’m investigating. Missing doc sections likely cause this, so we probably want to add some placeholder.
- Tags do not look as in design. I tried implementing them correctly but didn't have enough time to finish everything. Some initial implementation is in place, though. I will create a separate task for them.

https://github.com/enso-org/enso/assets/6566674/a656ae78-5d4c-45f4-a0a5-e07fa573253e

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/gui2/src/assets/icon-examples.svg
Comment (vitvakatu) [inline app/gui2/src/assets/icon-examples.svg:1]: I’m adding additional svg files because masks and clipPath do not work with `<use>` tag which is used by `SvgIcon`. Besides, I need custom styling and I don’t want to abuse `SvgIcon` too much.

File: app/gui2/src/components/ComponentBrowser.vue
Comment (vitvakatu) [inline app/gui2/src/components/ComponentBrowser.vue:51]: Funny, I broke this thing when addressing review comments in my last PR.

File: app/gui2/src/components/ComponentBrowser.vue
Comment (vitvakatu) [inline app/gui2/src/components/ComponentBrowser.vue]: Stealed from #7993

File: app/gui2/src/components/DocumentationPanel.vue
Comment (vitvakatu) [inline app/gui2/src/components/DocumentationPanel.vue]: It's awkward, but I didn't think of a better way. I guess I need to change the definition of the `Docs` type somehow because the current one causes too many type acrobatics in the code. I’d prefer to do that in a separate PR, though.

File: app/gui2/src/components/DocumentationPanel.vue
Comment (vitvakatu) [inline app/gui2/src/components/DocumentationPanel.vue]: Is there some way to make it less painful with Typescript? In pure JS, `docs.Function.sections || docs.Type.sections || … || fallback` works, but Typescript does not like it. It has a point I guess, but it still sucks.

File: app/gui2/src/components/documentation/DocsExamples.vue
Comment (vitvakatu) [inline app/gui2/src/components/documentation/DocsExamples.vue]: Note to future self: we would need some validation to protect from XSS-attacks.

**Review by vitvakatu — COMMENTED** [review]
Some notes for reviewers, as my Typescript-fu sucks.

File: app/gui2/src/components/DocumentationPanel.vue
Comment (somebody1234) [reply app/gui2/src/components/DocumentationPanel.vue]: no - typescript doesn't like this kind of discriminated union in general - it only has dedicated support for unions discriminated by the value of a specific property: [more info](https://basarat.gitbook.io/typescript/type-system/discriminated-unions).
i believe the only way to make this work, would be to add `Function?: undefined, Type?: undefined` etc. to every variant.

iirc this is something i brought up when pawel introduced the initial AST API, but i guess the conclusion was, it's not worth risking breaking the existing API (and/or making the API inconsistent between languages) for slightly better ergonomics.

File: app/gui2/src/components/documentation/DocsSynopsis.vue
Comment (somebody1234) [inline app/gui2/src/components/documentation/DocsSynopsis.vue]: I believe we're going with `:is()` instead of `:deep()`? Pinging @Frizi for confirmation. Not sure whether there's a meaningful difference apart from `:is` being a relatively new feature. ([caniuse](https://caniuse.com/?search=is()) - Chromium 88+, Safari 14+, FF 78+)

File: app/gui2/src/assets/icon-examples.svg
Comment (Frizi) [reply app/gui2/src/assets/icon-examples.svg:1]: My goal is to eventually have a single-button export from figma to a single "all icons" file. The separation of those icons means it's not possible, so I'd really like to remain consistent and find a way to use the existing convention.

As those specific icons go, It should be very easy to remove the clip path usages from them. But ideally we would find a way to support them instead.

File: app/gui2/src/components/documentation/DocsList.vue
Comment (Frizi) [inline app/gui2/src/components/documentation/DocsList.vue]: Make that a `firstSynopsisParagraph` computed value, so it can be reused.

File: app/gui2/src/components/DocumentationPanel.vue
Comment (Frizi) [inline app/gui2/src/components/DocumentationPanel.vue]: This whole type and corresponding functions should be moved out of the component, possibly to a `docParser` module. Exporting types from components is rarely a good idea.

Also, If I understand this correctly, all objects of `Docs` type are constructed manually in code (i.e. not from parser), so there is no point in using this unconventional type union format. It is better to use tagged unions that typescripts understands well. One nice way to do it is to [include a common "kind" property in all union interfaces](https://mariusschulz.com/blog/tagged-union-types-in-typescript#modeling-payment-methods-with-tagged-union-types). You could even make an `enum` for the available kinds, similarily to `SuggestionKind`, but it's not strictly necesasry. In general, that should remove a lot of type gymnastics that you had to do in the following code.

File: app/gui2/src/components/DocumentationPanel.vue
Comment (Frizi) [inline app/gui2/src/components/DocumentationPanel.vue]: The stores in general should be accessed at the module level, not inside a computed value.

File: app/gui2/src/components/DocumentationPanel.vue
Comment (Frizi) [inline app/gui2/src/components/DocumentationPanel.vue]: Default imports have a shorter preferred syntax, like
```ts
import DocsExamples from '@/components/documentation/DocsExamples.vue'
```

File: app/gui2/src/components/DocumentationPanel.vue
Comment (Frizi) [inline app/gui2/src/components/DocumentationPanel.vue]: I'm not a fan of this. The styles are scoped for a reason. Is the `deep` really necessary here?

Also, this probably should be a margin. Margins are usually logically considered to be "managed by the parent elements". CSS kinda took that more right with the grid `gap` property. Here, I would either use `gap`s, or at least move the class declaration to the component that styles it. Note that vue [allows declaring classes or other properties on components](https://vuejs.org/guide/essentials/class-and-style.html#with-components) in a template. Also, since we follow the convention of giving all component root nodes a class that matches its name, you can also use it, treating it kinda like a "tag selector", but for components (e.g. adding a style to all headers withing specific group with with `.someGroup .DocHeader` selector). This is possible, because the component root DOM nodes are considered to be both in their own scope, as well as in scope of their parent.

Interesting article for more context on CSS "property ownership": [**Axiomatic CSS and Lobotomized Owls**](https://alistapart.com/article/axiomatic-css-and-lobotomized-owls/)
I've used that technique previously, and it works pretty well for global styles, but isn't that great with vue scopes (which are just [fancy CSS selector macros](https://vuejs.org/api/sfc-css-features.html#scoped-css)). Now we should probably use `gap`s for similar behavior. But the idea of "who really owns that property" still stands.

File: app/gui2/src/components/documentation/DocsHeader.vue
Comment (Frizi) [inline app/gui2/src/components/documentation/DocsHeader.vue]:
```suggestion
const classes = computed(() => ['Header', props.kind])
```

File: app/gui2/src/components/documentation/DocsHeader.vue
Comment (Frizi) [inline app/gui2/src/components/documentation/DocsHeader.vue]: Avoid unnecessary CSS selector nesting. If something has its own class and the style always applies, just make it a top-level style. We have scopes after all.

File: app/gui2/src/components/documentation/DocsSynopsis.vue
Comment (Frizi) [inline app/gui2/src/components/documentation/DocsSynopsis.vue]: The `??` operator in templates makes syntax highlighting break. Here it would be fine to use `||` instead.

File: app/gui2/src/components/documentation/DocsSynopsis.vue
Comment (Frizi) [inline app/gui2/src/components/documentation/DocsSynopsis.vue]: Each independent `v-if` that ends up evaluating to `false` leaves a DOM comment node behind. If you need to select one of many options in the template, use [`v-else-if`](https://vuejs.org/guide/essentials/conditional.html#v-else-if). Vue then uses the fact that only one of many nodes can exist at a time, and doesn't need to add those placeholder comments.

**Review by Frizi — CHANGES_REQUESTED** [review]

File: app/gui2/src/assets/icon-examples.svg
Comment (vitvakatu) [reply app/gui2/src/assets/icon-examples.svg:1]: I don’t have write access to the Figma design, so I can‘t modify these icons myself. Besides, https://github.com/enso-org/enso/pull/7945 has improvements for the `SvgIcon`, which should allow me to use it in the future.

So I suggest doing it separately.

File: app/gui2/src/components/documentation/DocsList.vue
Comment (vitvakatu) [reply app/gui2/src/components/documentation/DocsList.vue]: It can’t be a computed property directly because we’re inside the `v-for` loop. I can precompute an array of annotations in advance, though. (did it)

File: app/gui2/src/components/DocumentationPanel.vue
Comment (vitvakatu) [reply app/gui2/src/components/DocumentationPanel.vue]: Types and supporting functions were moved to a separate module using the suggested approach. The code indeed became much more straightforward!

File: app/gui2/src/components/DocumentationPanel.vue
Comment (vitvakatu) [reply app/gui2/src/components/DocumentationPanel.vue]: Fixed

File: app/gui2/src/components/DocumentationPanel.vue
Comment (vitvakatu) [reply app/gui2/src/components/DocumentationPanel.vue]: Fixed

File: app/gui2/src/components/DocumentationPanel.vue
Comment (vitvakatu) [reply app/gui2/src/components/DocumentationPanel.vue]: I removed this particular instance of `:deep`, it was not needed after a refactoring. Thanks for an interesting read, I don‘t think it should be applied here. I refactored the layout a bit and managed to achieve the needed result much easier with a single additional div in DocumentationPanel.

File: app/gui2/src/components/documentation/DocsHeader.vue
Comment (vitvakatu) [reply app/gui2/src/components/documentation/DocsHeader.vue]: Refactored, and removed nesting in places where it was used for readability only.

File: app/gui2/src/components/documentation/DocsSynopsis.vue
Comment (vitvakatu) [reply app/gui2/src/components/documentation/DocsSynopsis.vue]: Fixed

File: app/gui2/src/components/documentation/DocsSynopsis.vue
Comment (vitvakatu) [reply app/gui2/src/components/documentation/DocsSynopsis.vue]: Done

**Review by Frizi — APPROVED** [review]
The only thing that I still am not a fan is the separate icon files. But I think it can be merged as is and fixed up later when we work on icon update workflow. That should be fairly straightforward.

**Pull request merged**
